main/src/ca/uwaterloo/flix/language/phase/Parser.scala

```diff
@@ -214,28 +214,42 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
 
     def Class: Rule1[ParsedAst.Declaration] = {
       def MarkerClass: Rule1[ParsedAst.Declaration.Class] = rule {
-        Documentation ~ Modifiers ~ SP ~ keyword("trait") ~ WS ~ Names.Class ~ optWS ~ TypeParams ~ push(Nil) ~ SP ~> ParsedAst.Declaration.Class
+        Documentation ~ Modifiers ~ SP ~ keyword("class") ~ WS ~ Names.Class ~ optWS ~ TypeParams ~ push(Nil) ~ SP ~> ParsedAst.Declaration.Class
       }
 
       def ClassWithSigs: Rule1[ParsedAst.Declaration.Class] = rule {
-        Documentation ~ Modifiers ~ SP ~ keyword("trait") ~ WS ~ Names.Class ~ optWS ~ TypeParams ~ optWS ~ "{" ~ optWS ~ zeroOrMore(Declarations.Sig).separatedBy(WS) ~ optWS ~ "}" ~ SP ~> ParsedAst.Declaration.Class
+        Documentation ~ Modifiers ~ SP ~ keyword("class") ~ WS ~ Names.Class ~ optWS ~ TypeParams ~ optWS ~ "{" ~ optWS ~ zeroOrMore(Declarations.Sig).separatedBy(WS) ~ optWS ~ "}" ~ SP ~> ParsedAst.Declaration.Class
       }
 
       rule {
         ClassWithSigs | MarkerClass
       }
     }
 
+    def Instance: Rule1[ParsedAst.Declaration] = {
+      def MarkerInstance:  Rule1[ParsedAst.Declaration.Instance] = rule {
+        Documentation ~ Modifiers ~ SP ~ keyword("instance") ~ WS ~ Names.QualifiedClass~ optWS ~ "[" ~ optWS ~ Type ~ optWS ~ "]" ~ push(Nil) ~ SP ~> ParsedAst.Declaration.Instance
+      }
+
+      def InstanceWithDefs: Rule1[ParsedAst.Declaration.Instance] = rule {
+        Documentation ~ Modifiers ~ SP ~ keyword("instance") ~ WS ~ Names.QualifiedClass ~ optWS ~ "[" ~ optWS ~ Type ~ optWS ~ "]" ~ optWS ~ "{" ~ optWS ~ zeroOrMore(Declarations.Def).separatedBy(WS) ~ optWS ~ "}" ~ SP ~> ParsedAst.Declaration.Instance
+      }
+
+      rule {
```

Comment:
In the interest of parsing performance, it is best not to have rules that share long prefixes.

I would add two rules `EmptyBody` and `NonEmptyBody` (or something like that).